File: Backend/app/rendering/midi_parser.py
"""MIDI file -> `ParsedMIDI`.

Ported from the iOS app's `MIDIParser.swift` (root `plan.md` M2) using
`mido` instead of AudioToolbox's `MusicSequence` — same voice-part heuristic
(track-name matching, falling back to mean-pitch ranking), same tempo-map-
aware tick->ms conversion, same SMF format-0-vs-1 handling.
"""

from __future__ import annotations

from collections import defaultdict, deque
from pathlib import Path

import mido

from app.rendering.models import (
    BackingNote,
    MIDILyricEvent,
    MIDINote,
    MIDITimeSignature,
    ParsedMIDI,
    VoicePart,
)

# Full-word aliases per part, matched as a *prefix* of the (trimmed,
# lowercased) track name — covers real-world split-part naming like
# "Soprano 1", "Soprano II", "Altos", "Tenor 2", "Sop.", "Alt".
_WORD_ALIASES: dict[VoicePart, list[str]] = {
    VoicePart.soprano: ["soprano", "sop"],
    VoicePart.alto: ["alto", "alt"],
    VoicePart.tenor: ["tenor", "ten"],
    VoicePart.bass: ["bass", "bs"],
}

# Single-letter shorthand per part ("S", "S1", "S 2", "S.", "T II", "B2") —
# matched only when the *entire* name reduces to the code plus
# punctuation/numbering, since a bare letter as a prefix elsewhere would
# false-positive too easily (e.g. "Strings").
_SHORT_CODES: dict[str, VoicePart] = {"s": VoicePart.soprano, "a": VoicePart.alto, "t": VoicePart.tenor, "b": VoicePart.bass}
_SHORT_CODE_TRAILER = set(".0123456789ivx ")

# Key-signature name (mido's `key_signature` meta message `.key`, e.g. "C",
# "F#m") -> signed fifths count, matching the MIDI key-signature meta-event's
# `sf` byte semantics the Swift parser reads directly.
_KEY_TO_FIFTHS: dict[str, int] = {
    "C": 0, "G": 1, "D": 2, "A": 3, "E": 4, "B": 5, "F#": 6, "C#": 7,
    "F": -1, "Bb": -2, "Eb": -3, "Ab": -4, "Db": -5, "Gb": -6, "Cb": -7,
    "Am": 0, "Em": 1, "Bm": 2, "F#m": 3, "C#m": 4, "G#m": 5, "D#m": 6, "A#m": 7,
    "Dm": -1, "Gm": -2, "Cm": -3, "Fm": -4, "Bbm": -5, "Ebm": -6, "Abm": -7,
}


class MIDIParserError(Exception):
    pass


def _match_voice_part(raw_name: str) -> VoicePart | None:
    name = raw_name.strip().lower()
    if not name:
        return None

    for part, aliases in _WORD_ALIASES.items():
        if any(name.startswith(alias) for alias in aliases):
            return part

    code, rest = name[0], name[1:]
    if not all(c in _SHORT_CODE_TRAILER for c in rest):
        return None
    return _SHORT_CODES.get(code)


def _assign_voice_parts(tracks: list[tuple[str | None, list[float]]]) -> dict[int, VoicePart]:
    """Maps track index -> voice part. Tracks that can't be confidently
    mapped (accompaniment, unnamed extras, an ambiguous leftover count) are
    simply absent from the result."""
    assignments: dict[int, VoicePart] = {}
    assigned_parts: set[VoicePart] = set()

    # Pass 1: track-name meta-events. Multiple tracks can map to the same
    # part on purpose — divisi splits ("Soprano 1"/"Soprano 2") both belong
    # in the same voice-part bucket.
    for index, (name, _pitches) in enumerate(tracks):
        if name is None:
            continue
        part = _match_voice_part(name)
        if part is None:
            continue
        assignments[index] = part
        assigned_parts.add(part)

    # Pass 2: mean-pitch fallback (high->low) for whatever voice parts are
    # still unassigned, drawn only from tracks that have notes and weren't
    # already name-matched to a different part.
    remaining_parts = [p for p in VoicePart if p not in assigned_parts]
    if not remaining_parts:
        return assignments

    candidates = [i for i, (_name, pitches) in enumerate(tracks) if i not in assignments and pitches]
    if len(candidates) != len(remaining_parts):
        # More or fewer note-bearing candidate tracks than remaining voice
        # parts — too ambiguous to guess at safely.
        return assignments

    ranked = sorted(candidates, key=lambda i: sum(tracks[i][1]) / len(tracks[i][1]), reverse=True)
    for part, index in zip(remaining_parts, ranked):
        assignments[index] = part
    return assignments


def _is_format_zero(file_path: Path) -> bool:
    """Reads the raw MThd header directly (format field), same as the Swift
    parser — mido's `MidiFile.type` would also tell us this, but the intent
    (and prior-art doc comment) is preserved by reading the header field
    directly rather than trusting a library's higher-level interpretation."""
    try:
        with open(file_path, "rb") as handle:
            header = handle.read(10)
    except OSError:
        return False
    if len(header) != 10 or header[:4] != b"MThd":
        return False
    fmt = (header[8] << 8) | header[9]
    return fmt == 0


def _absolute_ticks(track: mido.MidiTrack) -> list[tuple[int, mido.Message]]:
    events: list[tuple[int, mido.Message]] = []
    t = 0
    for msg in track:
        t += msg.time
        events.append((t, msg))
    return events


def _split_by_channel(track: mido.MidiTrack) -> dict[int, list[tuple[int, mido.Message]]]:
    """Format-0 files put every voice in one track, distinguished only by
    MIDI channel — split into one virtual track per channel, mirroring the
    Swift parser's use of `.smf_ChannelsToTracks` for format-0 files. Meta
    events (name/lyric/etc.) have no channel and aren't attributed to any
    single voice here, same as that load flag doesn't preserve per-channel
    names either — the mean-pitch fallback covers this case."""
    by_channel: dict[int, list[tuple[int, mido.Message]]] = defaultdict(list)
    t = 0
    for msg in track:
        t += msg.time
        channel = getattr(msg, "channel", None)
        if channel is not None:
            by_channel[channel].append((t, msg))
    return dict(by_channel)


def _build_tempo_map(all_events: list[tuple[int, mido.Message]]) -> list[tuple[int, int]]:
    """Returns [(abs_tick, microseconds_per_beat), ...] sorted by tick,
    always starting with a tick-0 entry (default 500000 usec/beat = 120bpm
    if the file has none) — needed to convert ticks to real seconds across
    tempo changes, same as the Swift parser reading the dedicated tempo
    track via `MusicSequenceGetSecondsForBeats`."""
    changes = sorted(
        {(tick, msg.tempo) for tick, msg in all_events if msg.type == "set_tempo"},
        key=lambda pair: pair[0],
    )
    if not changes or changes[0][0] != 0:
        changes.insert(0, (0, 500000))
    # Collapse duplicate ticks (keep the last), preserving order.
    merged: list[tuple[int, int]] = []
    for tick, tempo in changes:
        if merged and merged[-1][0] == tick:
            merged[-1] = (tick, tempo)
        else:
            merged.append((tick, tempo))
    return merged


def _ticks_to_ms(tempo_map: list[tuple[int, int]], ticks_per_beat: int, tick: int) -> int:
    """Integrates real elapsed time across every tempo-map segment up to
    `tick`, so a tempo change mid-note (or mid-piece) doesn't skew timing —
    matching the Swift parser's per-note recomputation via the tempo track
    rather than assuming one constant tempo throughout."""
    elapsed_us = 0.0
    for i, (seg_start, tempo) in enumerate(tempo_map):
        seg_end = tempo_map[i + 1][0] if i + 1 < len(tempo_map) else None
        if tick <= seg_start:
            break
        segment_ticks = tick - seg_start if seg_end is None else min(tick, seg_end) - seg_start
        elapsed_us += segment_ticks * tempo / ticks_per_beat
        if seg_end is not None and tick <= seg_end:
            break
    return round(elapsed_us / 1000)


def parse_midi(file_path: str | Path) -> ParsedMIDI:
    file_path = Path(file_path)
    try:
        mid = mido.MidiFile(str(file_path))
    except (OSError, EOFError, ValueError) as exc:
        raise MIDIParserError(f"Failed to load MIDI file: {exc}") from exc

    ticks_per_beat = mid.ticks_per_beat or 480

    if mid.type == 0 or _is_format_zero(file_path):
        raw_tracks_events = list(_split_by_channel(mid.tracks[0]).values()) if mid.tracks else []
    else:
        raw_tracks_events = [_absolute_ticks(track) for track in mid.tracks]

    all_events = [pair for track in raw_tracks_events for pair in track]
    tempo_map = _build_tempo_map(all_events)

    def to_ms(tick: int) -> int:
        return _ticks_to_ms(tempo_map, ticks_per_beat, tick)

    time_signature = MIDITimeSignature.default()
    for _tick, msg in sorted(all_events, key=lambda pair: pair[0]):
        if msg.type == "time_signature":
            time_signature = MIDITimeSignature(numerator=msg.numerator, denominator=msg.denominator)
            break

    tempo_bpm = 60_000_000 / tempo_map[0][1]

    # Per-track name / lyrics / key-signature / note pairing.
    names: list[str | None] = []
    track_lyrics: list[list[tuple[int, str]]] = []
    track_notes: list[list[tuple[int, int, int]]] = []  # (start_tick, duration_ticks, pitch)
    key_signature_fifths: int | None = None

    for events in raw_tracks_events:
        name: str | None = None
        lyrics: list[tuple[int, str]] = []
        notes: list[tuple[int, int, int]] = []
        pending: dict[int, deque[int]] = defaultdict(deque)  # pitch -> [start_tick, ...]

        for tick, msg in events:
            if msg.type == "track_name" and name is None:
                name = msg.name
            elif msg.type == "lyrics":
                lyrics.append((tick, msg.text))
            elif msg.type == "key_signature" and key_signature_fifths is None:
                key_signature_fifths = _KEY_TO_FIFTHS.get(msg.key, 0)
            elif msg.type == "note_on" and msg.velocity > 0:
                pending[msg.note].append(tick)
            elif msg.type in ("note_off", "note_on"):  # note_on velocity==0 == note_off
                starts = pending.get(msg.note)
                if starts:
                    start_tick = starts.popleft()
                    notes.append((start_tick, tick - start_tick, msg.note))

        notes.sort(key=lambda n: n[0])
        names.append(name)
        track_lyrics.append(lyrics)
        track_notes.append(notes)

    assignments = _assign_voice_parts(
        [(names[i], [float(p) for _s, _d, p in track_notes[i]]) for i in range(len(raw_tracks_events))]
    )

    notes: list[MIDINote] = []
    lyrics: list[MIDILyricEvent] = []
    backing_notes: list[BackingNote] = []
    max_tick = 0

    for index in range(len(raw_tracks_events)):
        voice_part = assignments.get(index)
        for start_tick, duration_ticks, pitch in track_notes[index]:
            start_ms = to_ms(start_tick)
            end_ms = to_ms(start_tick + duration_ticks)
            duration_ms = max(0, end_ms - start_ms)
            if voice_part is not None:
                notes.append(MIDINote(pitch=pitch, start_ms=start_ms, duration_ms=duration_ms, voice_part=voice_part))
            else:
                backing_notes.append(BackingNote(pitch=pitch, start_ms=start_ms, duration_ms=duration_ms))
            max_tick = max(max_tick, start_tick + duration_ticks)
        if voice_part is not None:
            for tick, text in track_lyrics[index]:
                lyrics.append(MIDILyricEvent(text=text, time_ms=to_ms(tick), voice_part=voice_part))

    notes.sort(key=lambda n: n.start_ms)
    lyrics.sort(key=lambda l: l.time_ms)
    backing_notes.sort(key=lambda n: n.start_ms)

    return ParsedMIDI(
        notes=notes,
        lyrics=lyrics,
        tempo_bpm=tempo_bpm,
        time_signature=time_signature,
        key_signature_fifths=key_signature_fifths or 0,
        track_voice_parts=assignments,
        backing_notes=backing_notes,
        duration_ms=to_ms(max_tick),
    )

File: DivisiKit/MIDIParser.swift
import AudioToolbox
import Foundation

enum MIDIParserError: Error {
    case sequenceCreationFailed(OSStatus)
    case fileLoadFailed(OSStatus)
    case trackEnumerationFailed(OSStatus)
}

/// Parses a Standard MIDI File into notes + lyric events, resolved to a
/// voice part.
///
/// Uses AudioToolbox's `MusicSequence`/`MusicEventIterator` C API rather
/// than the newer `AVAudioSequencer`/`AVMusicTrack.enumerateEvents`: the
/// modern `AVMIDIMetaEvent` only exposes `.type`, not the event's raw
/// bytes, so there's no supported way to read lyric or track-name text back
/// through it. `MusicEventIterator` exposes the full `MIDIMetaEvent` struct
/// (type + raw data), which is what that extraction actually needs. Both
/// APIs sit on the same underlying engine — this is still AVFoundation's
/// own MIDI parsing, one layer down.
final class MIDIParser {

    private static let metaEventTypeTrackName: UInt8 = 0x03
    private static let metaEventTypeLyric: UInt8 = 0x05
    private static let metaEventTypeTimeSignature: UInt8 = 0x58
    private static let metaEventTypeKeySignature: UInt8 = 0x59

    private struct RawTrack {
        let name: String?
        let notes: [(beat: MusicTimeStamp, durationBeats: Double, pitch: UInt8)]
        let lyrics: [(beat: MusicTimeStamp, text: String)]
        let keySignatureFifths: Int?
    }

    func parse(fileURL: URL) throws -> ParsedMIDI {
        var sequence: MusicSequence?
        var status = NewMusicSequence(&sequence)
        guard status == noErr, let sequence else {
            throw MIDIParserError.sequenceCreationFailed(status)
        }
        defer { DisposeMusicSequence(sequence) }

        // Real-world choir exports show up as either SMF format 1 (one
        // track per voice — what our fixtures use, and the common case for
        // a DAW/notation-software export) or format 0 (everything flattened
        // into a single track, voices distinguished only by MIDI channel).
        // Loading a format-1 file with `.smf_ChannelsToTracks` would merge
        // same-channel tracks together and destroy the voice split; loading
        // a format-0 file with `.smf_PreserveTracks` leaves every voice
        // stuck in one track. Pick the flag the file's own header asks for.
        let loadFlags: MusicSequenceLoadFlags = Self.isFormatZero(fileURL) ? .smf_ChannelsToTracks : .smf_PreserveTracks
        status = MusicSequenceFileLoad(sequence, fileURL as CFURL, .midiType, loadFlags)
        guard status == noErr else {
            throw MIDIParserError.fileLoadFailed(status)
        }

        var trackCount: UInt32 = 0
        status = MusicSequenceGetTrackCount(sequence, &trackCount)
        guard status == noErr else {
            throw MIDIParserError.trackEnumerationFailed(status)
        }

        var rawTracks: [RawTrack] = []
        for i in 0..<trackCount {
            var track: MusicTrack?
            guard MusicSequenceGetIndTrack(sequence, i, &track) == noErr, let track else { continue }
            rawTracks.append(readTrack(track))
        }

        let assignments = Self.assignVoiceParts(rawTracks.map { (name: $0.name, pitches: $0.notes.map { Double($0.pitch) }) })

        var notes: [MIDINote] = []
        var lyrics: [MIDILyricEvent] = []
        for (index, raw) in rawTracks.enumerated() {
            guard let voicePart = assignments[index] else { continue }
            for note in raw.notes {
                let startMs = secondsToMs(secondsForBeats(sequence, note.beat))
                // Recompute the end from the tempo map too, rather than
                // assuming a constant tempo across the note's duration — a
                // tempo change mid-note would otherwise skew durationMs.
                let endMs = secondsToMs(secondsForBeats(sequence, note.beat + note.durationBeats))
                notes.append(MIDINote(pitch: note.pitch, startMs: startMs, durationMs: max(0, endMs - startMs), voicePart: voicePart))
            }
            for lyric in raw.lyrics {
                lyrics.append(MIDILyricEvent(text: lyric.text, timeMs: secondsToMs(secondsForBeats(sequence, lyric.beat)), voicePart: voicePart))
            }
        }

        notes.sort { $0.startMs < $1.startMs }
        lyrics.sort { $0.timeMs < $1.timeMs }

        // Key signature: first occurrence wins, in track order — mid-file
        // changes aren't tracked, see `ParsedMIDI`'s doc comment.
        let keySignatureFifths = rawTracks.compactMap(\.keySignatureFifths).first ?? 0
        let (tempoBPM, timeSignature) = Self.readTempoTrack(sequence)

        return ParsedMIDI(
            notes: notes,
            lyrics: lyrics,
            tempoBPM: tempoBPM,
            timeSignature: timeSignature,
            keySignatureFifths: keySignatureFifths,
            trackVoiceParts: assignments
        )
    }

    /// Tempo (0x51) and time-signature (0x58) meta-events don't come back
    /// through a regular track's `MusicEventIterator`, even when the source
    /// SMF interleaved them into a voice track rather than a dedicated
    /// conductor track (real-world files do this) — `MusicSequenceFileLoad`
    /// pulls both into its own dedicated tempo track instead. Verified
    /// against a real notation-software export where both meta types were
    /// silently missing from every regular track but present here.
    private static func readTempoTrack(_ sequence: MusicSequence) -> (bpm: Double, timeSignature: MIDITimeSignature) {
        var tempoTrack: MusicTrack?
        guard MusicSequenceGetTempoTrack(sequence, &tempoTrack) == noErr, let tempoTrack else {
            return (120, .defaultSignature)
        }
        var iterator: MusicEventIterator?
        guard NewMusicEventIterator(tempoTrack, &iterator) == noErr, let iterator else {
            return (120, .defaultSignature)
        }
        defer { DisposeMusicEventIterator(iterator) }

        var bpm: Double?
        var timeSignature: MIDITimeSignature?
        var hasEvent: DarwinBoolean = false
        MusicEventIteratorHasCurrentEvent(iterator, &hasEvent)
        while hasEvent.boolValue, bpm == nil || timeSignature == nil {
            var timeStamp: MusicTimeStamp = 0
            var eventType: MusicEventType = 0
            var eventData: UnsafeRawPointer?
            var eventDataSize: UInt32 = 0
            let status = MusicEventIteratorGetEventInfo(iterator, &timeStamp, &eventType, &eventData, &eventDataSize)
            if status == noErr, let eventData {
                if eventType == kMusicEventType_ExtendedTempo, bpm == nil {
                    let tempo = eventData.assumingMemoryBound(to: ExtendedTempoEvent.self).pointee
                    if tempo.bpm > 0 { bpm = tempo.bpm }
                } else if eventType == kMusicEventType_Meta, timeSignature == nil {
                    let meta = eventData.assumingMemoryBound(to: MIDIMetaEvent.self).pointee
                    if meta.metaEventType == Self.metaEventTypeTimeSignature, meta.dataLength >= 2 {
                        let dataOffset = MemoryLayout<MIDIMetaEvent>.offset(of: \.data)!
                        let bytes = [UInt8](UnsafeRawBufferPointer(start: UnsafeRawPointer(eventData).advanced(by: dataOffset), count: Int(meta.dataLength)))
                        timeSignature = MIDITimeSignature(numerator: Int(bytes[0]), denominator: 1 << Int(bytes[1]))
                    }
                }
            }
            MusicEventIteratorNextEvent(iterator)
            MusicEventIteratorHasCurrentEvent(iterator, &hasEvent)
        }
        return (bpm ?? 120, timeSignature ?? .defaultSignature)
    }

    // MARK: - Track reading

    private func readTrack(_ track: MusicTrack) -> RawTrack {
        var name: String?
        var notes: [(beat: MusicTimeStamp, durationBeats: Double, pitch: UInt8)] = []
        var lyrics: [(beat: MusicTimeStamp, text: String)] = []
        var keySignatureFifths: Int?

        var iterator: MusicEventIterator?
        guard NewMusicEventIterator(track, &iterator) == noErr, let iterator else {
            return RawTrack(name: nil, notes: [], lyrics: [], keySignatureFifths: nil)
        }
        defer { DisposeMusicEventIterator(iterator) }

        var hasEvent: DarwinBoolean = false
        MusicEventIteratorHasCurrentEvent(iterator, &hasEvent)
        while hasEvent.boolValue {
            var timeStamp: MusicTimeStamp = 0
            var eventType: MusicEventType = 0
            var eventData: UnsafeRawPointer?
            var eventDataSize: UInt32 = 0
            let status = MusicEventIteratorGetEventInfo(iterator, &timeStamp, &eventType, &eventData, &eventDataSize)
            if status == noErr, let eventData {
                switch eventType {
                case kMusicEventType_MIDINoteMessage:
                    let note = eventData.assumingMemoryBound(to: MIDINoteMessage.self).pointee
                    notes.append((beat: timeStamp, durationBeats: Double(note.duration), pitch: note.note))
                case kMusicEventType_Meta:
                    let meta = eventData.assumingMemoryBound(to: MIDIMetaEvent.self).pointee
                    let dataOffset = MemoryLayout<MIDIMetaEvent>.offset(of: \.data)!
                    let bytes = [UInt8](UnsafeRawBufferPointer(start: UnsafeRawPointer(eventData).advanced(by: dataOffset), count: Int(meta.dataLength)))
                    switch meta.metaEventType {
                    case Self.metaEventTypeTrackName:
                        name = String(bytes: bytes, encoding: .utf8)
                    case Self.metaEventTypeLyric:
                        if let text = String(bytes: bytes, encoding: .utf8) {
                            lyrics.append((beat: timeStamp, text: text))
                        }
                    case Self.metaEventTypeKeySignature where bytes.count >= 1:
                        // Signed byte: sharps if positive, flats if negative.
                        keySignatureFifths = Int(Int8(bitPattern: bytes[0]))
                    default:
                        break
                    }
                default:
                    break
                }
            }
            MusicEventIteratorNextEvent(iterator)
            MusicEventIteratorHasCurrentEvent(iterator, &hasEvent)
        }

        return RawTrack(name: name, notes: notes, lyrics: lyrics, keySignatureFifths: keySignatureFifths)
    }

    // MARK: - Format detection

    /// Reads the raw MThd header (fixed 14-byte chunk: "MThd" + 4-byte
    /// length + 2-byte format + 2-byte track count + 2-byte division) to
    /// get the SMF format field. This is the one place this parser reads
    /// raw file bytes directly — a fixed-offset fixed-size field, not the
    /// variable-length event stream `MusicSequence` already handles.
    /// Defaults to "not format 0" on any read/parse failure so a malformed
    /// header doesn't change behavior here — `MusicSequenceFileLoad` will
    /// surface the real error.
    private static func isFormatZero(_ fileURL: URL) -> Bool {
        guard let handle = try? FileHandle(forReadingFrom: fileURL) else { return false }
        defer { try? handle.close() }
        guard let data = try? handle.read(upToCount: 10), data.count == 10 else { return false }
        let header = [UInt8](data)
        guard header.prefix(4).elementsEqual("MThd".utf8) else { return false }
        let format = UInt16(header[8]) << 8 | UInt16(header[9])
        return format == 0
    }

    // MARK: - Voice-part mapping

    /// Full-word aliases per part, matched as a *prefix* of the (trimmed,
    /// lowercased) track name — covers real-world split-part naming like
    /// "Soprano 1", "Soprano II", "Altos", "Tenor 2", "Sop.", "Alt".
    private static let wordAliases: [VoicePart: [String]] = [
        .soprano: ["soprano", "sop"],
        .alto: ["alto", "alt"],
        .tenor: ["tenor", "ten"],
        .bass: ["bass", "bs"],
    ]

    /// Single-letter shorthand per part ("S", "S1", "S 2", "S.", "T II",
    /// "B2") — matched only when the *entire* name reduces to the code plus
    /// punctuation/numbering, since a bare letter as a prefix elsewhere
    /// would false-positive too easily (e.g. "Strings").
    private static let shortCodes: [Character: VoicePart] = ["s": .soprano, "a": .alto, "t": .tenor, "b": .bass]

    private static func matchVoicePart(forTrackName rawName: String) -> VoicePart? {
        let name = rawName.trimmingCharacters(in: .whitespaces).lowercased()
        guard !name.isEmpty else { return nil }

        if let part = wordAliases.first(where: { alias in alias.value.contains(where: name.hasPrefix) })?.key {
            return part
        }

        guard let code = name.first else { return nil }
        let rest = name.dropFirst()
        guard rest.allSatisfy({ ".0123456789ivx ".contains($0) }) else { return nil }
        return shortCodes[code]
    }

    /// Maps track index -> voice part. Tracks that can't be confidently
    /// mapped (accompaniment, unnamed extras, an ambiguous leftover count)
    /// are simply absent from the result — see the human spot-check task
    /// in the M2 plan for validating this against real files.
    private static func assignVoiceParts(_ tracks: [(name: String?, pitches: [Double])]) -> [Int: VoicePart] {
        var assignments: [Int: VoicePart] = [:]
        var assignedParts: Set<VoicePart> = []

        // Pass 1: track-name meta-events. Multiple tracks can map to the
        // same part on purpose — divisi splits ("Soprano 1"/"Soprano 2")
        // both belong in the same voice-part bucket.
        for (index, track) in tracks.enumerated() {
            guard let name = track.name, let part = matchVoicePart(forTrackName: name) else { continue }
            assignments[index] = part
            assignedParts.insert(part)
        }

        // Pass 2: mean-pitch fallback (high→low) for whatever voice parts
        // are still unassigned, drawn only from tracks that have notes and
        // weren't already name-matched to a different part.
        let remainingParts = VoicePart.allCases.filter { !assignedParts.contains($0) }
        guard !remainingParts.isEmpty else { return assignments }

        let candidates = tracks.indices.filter { assignments[$0] == nil && !tracks[$0].pitches.isEmpty }
        guard candidates.count == remainingParts.count else {
            // More or fewer note-bearing candidate tracks than remaining
            // voice parts — too ambiguous to guess at safely.
            return assignments
        }

        let ranked = candidates.sorted { meanPitch(tracks[$0].pitches) > meanPitch(tracks[$1].pitches) }
        for (part, index) in zip(remainingParts, ranked) {
            assignments[index] = part
        }
        return assignments
    }

    private static func meanPitch(_ pitches: [Double]) -> Double {
        pitches.reduce(0, +) / Double(pitches.count)
    }

    // MARK: - Time conversion

    private func secondsForBeats(_ sequence: MusicSequence, _ beats: MusicTimeStamp) -> Double {
        var seconds: Float64 = 0
        MusicSequenceGetSecondsForBeats(sequence, beats, &seconds)
        return seconds
    }

    private func secondsToMs(_ seconds: Double) -> Int {
        Int((seconds * 1000).rounded())
    }
}
